packages/notification-service/src/exchange/exchangeQuery.ts

```diff
@@ -0,0 +1,45 @@
+import { CURRENCY_ENUM } from '@celo/utils'
+import { ContractUtils } from '@celo/walletkit'
+import BigNumber from 'bignumber.js'
+import Web3 from 'web3'
+import { WEB3_PROVIDER_URL } from '../config'
+import { writeExchangeRatePair } from '../firebase'
+
+// Amounts to estimate the exchange rate, as the rate varies based on transaction size
+const SELL_AMOUNTS = {
+  [CURRENCY_ENUM.DOLLAR]: new BigNumber(10000 * 1000000000000000000), // 100 dollars
+  [CURRENCY_ENUM.GOLD]: new BigNumber(10 * 1000000000000000000), // 10 gold
+}
+
+export async function handleExchangeQuery(web3Instance: Web3) {
+  const fetchTime = Date.now().toString()
+  const [dollarMakerRate, goldMakerRate] = await Promise.all([
+    getExchangeRate(CURRENCY_ENUM.DOLLAR, web3Instance),
+    getExchangeRate(CURRENCY_ENUM.GOLD, web3Instance),
+  ])
+
+  writeExchangeRatePair(CURRENCY_ENUM.DOLLAR, dollarMakerRate.toString(), fetchTime)
+  writeExchangeRatePair(CURRENCY_ENUM.GOLD, goldMakerRate.toString(), fetchTime)
+}
+
+export async function getExchangeRate(makerToken: CURRENCY_ENUM, web3Instance: Web3) {
+  const rate = await ContractUtils.getExchangeRate(
+    web3Instance,
+    makerToken,
+    SELL_AMOUNTS[makerToken]
+  )
+  return rate
+}
+
+let web3: Web3
+export function getWeb3Instance(): Web3 {
+  if (web3) {
+    if (web3.eth.net.isListening()) {
```

Comment:
And if not?

Reply:
Else it continues to line 42 and creates a new web3 instance, as the connection has been lost

Reply:
I haven't been able to find a cleaner way than this nested if statement, as ` if (web3 && web3.eth.net.isListening())` would throw an error if web3 is not yet initialized

Reply:
Updated